File: packages/node/src/sdk/events/ai.ts
import { LogEvent, Severity } from '../../api/types';

/** Well-known AI / Kavia event names. */
export type AiEventName =
  | 'AI_PROMPT_SUBMITTED'
  | 'AI_RESPONSE_RECEIVED'
  | 'AI_TOOL_INVOCATION'
  | 'AI_MEMORY_RETRIEVAL'
  | 'AI_RAG_RETRIEVAL'
  | 'AI_AGENT_EXECUTION'
  | 'AI_SAFETY_FILTER_APPLIED';

/**
 * Caller-supplied fields for an AI structured event.
 */
export interface AiEventFields {
  message?: string;
  severity?: Severity | string;
  'kavia.provider'?: string;
  'kavia.model'?: string;
  'kavia.model.version'?: string;
  'kavia.prompt'?: string;
  'kavia.response'?: string;
  'kavia.tokens.prompt'?: number;
  'kavia.tokens.completion'?: number;
  'kavia.tokens.total'?: number;
  'kavia.latency.ms'?: number;
  'kavia.cost'?: number;
  'kavia.context_window.size'?: number;
  'kavia.memory.bytes'?: number;
  'kavia.conversation.id'?: string;
  'kavia.agent.id'?: string;
  'kavia.request.id'?: string;
  'kavia.finish_reason'?: string;
  'kavia.cache.hit'?: boolean;
  'kavia.temperature'?: number;
  'kavia.max_tokens'?: number;
  prompt?: string;
  response?: string;
  [key: string]: unknown;
}

const defaultMessages: Record<AiEventName, string> = {
  AI_PROMPT_SUBMITTED: 'AI prompt submitted',
  AI_RESPONSE_RECEIVED: 'AI response received',
  AI_TOOL_INVOCATION: 'AI tool invocation',
  AI_MEMORY_RETRIEVAL: 'AI memory retrieval',
  AI_RAG_RETRIEVAL: 'AI RAG retrieval',
  AI_AGENT_EXECUTION: 'AI agent execution',
  AI_SAFETY_FILTER_APPLIED: 'AI safety filter applied',
};

const kaviaKeys = [
  'kavia.provider',
  'kavia.model',
  'kavia.model.version',
  'kavia.prompt',
  'kavia.response',
  'kavia.tokens.prompt',
  'kavia.tokens.completion',
  'kavia.tokens.total',
  'kavia.latency.ms',
  'kavia.cost',
  'kavia.context_window.size',
  'kavia.memory.bytes',
  'kavia.conversation.id',
  'kavia.agent.id',
  'kavia.request.id',
  'kavia.finish_reason',
  'kavia.cache.hit',
  'kavia.temperature',
  'kavia.max_tokens',
] as const;

/**
 * Build a partial canonical LogEvent for an AI / Kavia interaction.
 *
 * @param eventName - AI lifecycle event name.
 * @param fields - Kavia attributes and optional message/severity overrides.
 * @returns A partial LogEvent suitable for `Logger.event` / `EventEmitter.emit`.
 */
export function buildAiEvent(
  eventName: AiEventName,
  fields: AiEventFields = {},
): Partial<LogEvent> {
  const { message, severity, prompt, response, ...rest } = fields;

  const event: Partial<LogEvent> = {
    ...rest,
    'event.name': eventName,
    message: message ?? defaultMessages[eventName],
    severity: severity ?? Severity.INFO,
  };

  if (prompt !== undefined && event['kavia.prompt'] === undefined) {
    event['kavia.prompt'] = prompt;
  }
  if (response !== undefined && event['kavia.response'] === undefined) {
    event['kavia.response'] = response;
  }

  for (const key of kaviaKeys) {
    if (fields[key] !== undefined) {
      event[key] = fields[key];
    }
  }

  return event;
}

/**
 * Build an AI request (prompt-side) event correlated by `kavia.request.id`.
 */
export function buildAiRequestEvent(fields: AiEventFields = {}): Partial<LogEvent> {
  return buildAiEvent('AI_PROMPT_SUBMITTED', {
    message: fields.message ?? 'AI request submitted',
    ...fields,
  });
}

/**
 * Build an AI response event correlated by `kavia.request.id`.
 */
export function buildAiResponseEvent(fields: AiEventFields = {}): Partial<LogEvent> {
  return buildAiEvent('AI_RESPONSE_RECEIVED', {
    message: fields.message ?? 'AI response received',
    ...fields,
  });
}

File: packages/python/dt3_sdk/events/ai.py
"""AI (kavia.*) event builders."""

from __future__ import annotations

from typing import Any, Dict, Optional


def build_ai_event(
    event_name: str,
    message: str,
    *,
    provider: Optional[str] = None,
    model: Optional[str] = None,
    model_version: Optional[str] = None,
    prompt: Optional[str] = None,
    response: Optional[str] = None,
    tokens_prompt: Optional[int] = None,
    tokens_completion: Optional[int] = None,
    tokens_total: Optional[int] = None,
    latency_ms: Optional[float] = None,
    cost: Optional[float] = None,
    context_window_size: Optional[int] = None,
    memory_bytes: Optional[int] = None,
    conversation_id: Optional[str] = None,
    agent_id: Optional[str] = None,
    request_id: Optional[str] = None,
    finish_reason: Optional[str] = None,
    cache_hit: Optional[bool] = None,
    temperature: Optional[float] = None,
    max_tokens: Optional[int] = None,
    severity: str = "INFO",
    **extra: Any,
) -> Dict[str, Any]:
    """Build a canonical AI LogEvent fragment using kavia.* attributes."""
    event: Dict[str, Any] = {
        "event.name": event_name,
        "message": message,
        "severity": severity,
    }
    mapping = {
        "kavia.provider": provider,
        "kavia.model": model,
        "kavia.model.version": model_version,
        "kavia.prompt": prompt,
        "kavia.response": response,
        "kavia.tokens.prompt": tokens_prompt,
        "kavia.tokens.completion": tokens_completion,
        "kavia.tokens.total": tokens_total,
        "kavia.latency.ms": latency_ms,
        "kavia.cost": cost,
        "kavia.context_window.size": context_window_size,
        "kavia.memory.bytes": memory_bytes,
        "kavia.conversation.id": conversation_id,
        "kavia.agent.id": agent_id,
        "kavia.request.id": request_id,
        "kavia.finish_reason": finish_reason,
        "kavia.cache.hit": cache_hit,
        "kavia.temperature": temperature,
        "kavia.max_tokens": max_tokens,
    }
    for key, value in mapping.items():
        if value is not None:
            event[key] = value
    event.update(extra)
    return event


def build_prompt_submitted(message: str = "AI prompt submitted", **kwargs: Any) -> Dict[str, Any]:
    return build_ai_event("AI_PROMPT_SUBMITTED", message, **kwargs)


def build_response_received(message: str = "AI response received", **kwargs: Any) -> Dict[str, Any]:
    return build_ai_event("AI_RESPONSE_RECEIVED", message, **kwargs)


def build_tool_invocation(message: str = "AI tool invocation", **kwargs: Any) -> Dict[str, Any]:
    return build_ai_event("AI_TOOL_INVOCATION", message, **kwargs)


def build_memory_retrieval(message: str = "AI memory retrieval", **kwargs: Any) -> Dict[str, Any]:
    return build_ai_event("AI_MEMORY_RETRIEVAL", message, **kwargs)


def build_rag_retrieval(message: str = "AI RAG retrieval", **kwargs: Any) -> Dict[str, Any]:
    return build_ai_event("AI_RAG_RETRIEVAL", message, **kwargs)


def build_agent_execution(message: str = "AI agent execution", **kwargs: Any) -> Dict[str, Any]:
    return build_ai_event("AI_AGENT_EXECUTION", message, **kwargs)


def build_safety_filter_applied(
    message: str = "AI safety filter applied", **kwargs: Any
) -> Dict[str, Any]:
    return build_ai_event("AI_SAFETY_FILTER_APPLIED", message, **kwargs)
